File: examples/cpp/backend_service.cpp
/**
 * @file backend_service.cpp
 * @brief iVIT-SDK Backend Engineer Service Example
 *
 * Target: Backend engineers who need to build stable AI inference services
 *         with monitoring and performance tracking.
 *
 * Features demonstrated:
 * - Service wrapper with monitoring
 * - FPS and latency tracking
 * - Thread-safe inference service pattern
 * - Statistics collection
 *
 * Note: Advanced callback system is available via Python API.
 *       C++ demonstrates manual monitoring patterns.
 *
 * Build:
 *   mkdir build && cd build
 *   cmake .. -DCMAKE_BUILD_TYPE=Release
 *   make backend_service
 *
 * Usage:
 *   ./backend_service <model_path>
 *   ./backend_service <model_path> --demo
 *   ./backend_service <model_path> --device cuda:0
 */

#include <iostream>
#include <iomanip>
#include <string>
#include <vector>
#include <chrono>
#include <atomic>
#include <mutex>
#include <numeric>
#include <deque>

#include <opencv2/opencv.hpp>

#include "ivit/ivit.hpp"

using namespace ivit;

/**
 * FPS Counter - Sliding window FPS calculation
 */
class FPSCounter {
public:
    explicit FPSCounter(size_t window_size = 30)
        : window_size_(window_size) {}

    void record(double latency_ms) {
        std::lock_guard<std::mutex> lock(mutex_);

        latencies_.push_back(latency_ms);

        while (latencies_.size() > window_size_) {
            latencies_.pop_front();
        }
    }

    double fps() const {
        std::lock_guard<std::mutex> lock(mutex_);
        if (latencies_.empty()) return 0.0;

        double total_ms = std::accumulate(latencies_.begin(), latencies_.end(), 0.0);
        double avg_ms = total_ms / latencies_.size();
        return avg_ms > 0 ? 1000.0 / avg_ms : 0.0;
    }

    double avg_latency_ms() const {
        std::lock_guard<std::mutex> lock(mutex_);
        if (latencies_.empty()) return 0.0;
        return std::accumulate(latencies_.begin(), latencies_.end(), 0.0) / latencies_.size();
    }

private:
    size_t window_size_;
    std::deque<double> latencies_;
    mutable std::mutex mutex_;
};

/**
 * Latency Logger - Tracks inference latency statistics
 */
class LatencyLogger {
public:
    void record(double latency_ms) {
        std::lock_guard<std::mutex> lock(mutex_);

        latencies_.push_back(latency_ms);
        total_latency_ += latency_ms;
        inference_count_++;

        // Keep last 1000 samples for rolling statistics
        if (latencies_.size() > 1000) {
            total_latency_ -= latencies_.front();
            latencies_.pop_front();
        }
    }

    double average_latency() const {
        std::lock_guard<std::mutex> lock(mutex_);
        if (latencies_.empty()) return 0.0;
        return total_latency_ / latencies_.size();
    }

    size_t count() const {
        std::lock_guard<std::mutex> lock(mutex_);
        return inference_count_;
    }

    double min_latency() const {
        std::lock_guard<std::mutex> lock(mutex_);
        if (latencies_.empty()) return 0.0;
        return *std::min_element(latencies_.begin(), latencies_.end());
    }

    double max_latency() const {
        std::lock_guard<std::mutex> lock(mutex_);
        if (latencies_.empty()) return 0.0;
        return *std::max_element(latencies_.begin(), latencies_.end());
    }

private:
    std::deque<double> latencies_;
    double total_latency_ = 0.0;
    size_t inference_count_ = 0;
    mutable std::mutex mutex_;
};

/**
 * Alert Monitor - Monitors for high latency events
 */
class AlertMonitor {
public:
    explicit AlertMonitor(double threshold_ms = 100.0)
        : threshold_ms_(threshold_ms), alert_count_(0) {}

    void check(double latency_ms) {
        if (latency_ms > threshold_ms_) {
            alert_count_++;
            std::cout << "[ALERT] High latency warning: " << std::fixed
                      << std::setprecision(1) << latency_ms << "ms "
                      << "(threshold: " << threshold_ms_ << "ms)" << std::endl;
        }
    }

    size_t alert_count() const { return alert_count_.load(); }
    double threshold() const { return threshold_ms_; }

private:
    double threshold_ms_;
    std::atomic<size_t> alert_count_;
};

/**
 * Inference Service - Wraps model with monitoring
 */
class InferenceService {
public:
    InferenceService(const std::string& model_path, const DeviceInfo& device)
        : fps_counter_(30),
          alert_monitor_(100.0) {

        // Load model using load_model API
        LoadConfig config;
        config.device = device.id;
        model_ = load_model(model_path, config);

        device_id_ = device.id;
        device_name_ = device.name;

        // Warmup
        warmup(10);
    }

    Results infer(const cv::Mat& image) {
        auto start = std::chrono::high_resolution_clock::now();

        auto results = model_->predict(image);

        auto end = std::chrono::high_resolution_clock::now();
        double latency = std::chrono::duration<double, std::milli>(end - start).count();

        // Record metrics
        fps_counter_.record(latency);
        latency_logger_.record(latency);
        alert_monitor_.check(latency);

        return results;
    }

    void warmup(int iterations = 10) {
        std::cout << "Running " << iterations << " warmup iterations..." << std::endl;

        cv::Mat dummy(480, 640, CV_8UC3);
        cv::randu(dummy, cv::Scalar(0, 0, 0), cv::Scalar(255, 255, 255));

        for (int i = 0; i < iterations; ++i) {
            model_->predict(dummy);
        }

        std::cout << "Warmup completed!" << std::endl;
    }

    // Statistics getters
    double fps() const { return fps_counter_.fps(); }
    double average_latency() const { return latency_logger_.average_latency(); }
    double min_latency() const { return latency_logger_.min_latency(); }
    double max_latency() const { return latency_logger_.max_latency(); }
    size_t inference_count() const { return latency_logger_.count(); }
    size_t alert_count() const { return alert_monitor_.alert_count(); }

    const std::string& device_id() const { return device_id_; }
    const std::string& device_name() const { return device_name_; }

private:
    std::shared_ptr<Model> model_;
    std::string device_id_;
    std::string device_name_;

    FPSCounter fps_counter_;
    LatencyLogger latency_logger_;
    AlertMonitor alert_monitor_;
};

void demonstrate_monitoring_patterns() {
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "Step 1: Monitoring Patterns" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    std::cout << "Monitoring components demonstrated:" << std::endl;
    std::cout << "  - FPSCounter: Sliding window FPS calculation" << std::endl;
    std::cout << "  - LatencyLogger: Latency statistics (min/max/avg)" << std::endl;
    std::cout << "  - AlertMonitor: High latency alerting" << std::endl;

    std::cout << "\nFor advanced callback system, use Python API:" << std::endl;
    std::cout << "  from ivit.core.callbacks import CallbackManager, FPSCounter" << std::endl;
    std::cout << "  manager = CallbackManager()" << std::endl;
    std::cout << "  manager.register('infer_end', FPSCounter())" << std::endl;
}

void demonstrate_cli_tools() {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Step 3: CLI Tools" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    std::cout << "Available CLI commands for backend engineers:" << std::endl;
    std::cout << std::endl;
    std::cout << "# System information" << std::endl;
    std::cout << "  ivit info" << std::endl;
    std::cout << std::endl;
    std::cout << "# List available devices" << std::endl;
    std::cout << "  ivit devices" << std::endl;
    std::cout << std::endl;
    std::cout << "# Performance benchmark" << std::endl;
    std::cout << "  ivit benchmark model.onnx --device cuda:0 --iterations 100" << std::endl;
    std::cout << std::endl;
    std::cout << "# Run inference" << std::endl;
    std::cout << "  ivit predict model.onnx image.jpg --output result.jpg" << std::endl;
    std::cout << std::endl;
    std::cout << "# Model conversion" << std::endl;
    std::cout << "  ivit convert model.onnx model.engine --format tensorrt --fp16" << std::endl;
    std::cout << std::endl;
    std::cout << "# Start REST API service" << std::endl;
    std::cout << "  ivit serve model.onnx --port 8080 --device cuda:0" << std::endl;
}

void run_demo(InferenceService& service) {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Step 2: Demo Inferences" << std::endl;
    std::cout << std::string(60, '=') << std::endl;

    // Create test image
    cv::Mat test_image(480, 640, CV_8UC3);
    cv::randu(test_image, cv::Scalar(0, 0, 0), cv::Scalar(255, 255, 255));

    std::cout << "Running 20 demo inferences..." << std::endl;

    for (int i = 0; i < 20; ++i) {
        auto results = service.infer(test_image);

        if ((i + 1) % 5 == 0) {
            std::cout << "  Completed " << (i + 1) << "/20, "
                      << "Current FPS: " << std::fixed << std::setprecision(1)
                      << service.fps() << std::endl;
        }
    }

    // Show statistics
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Statistics:" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Total inferences: " << service.inference_count() << std::endl;
    std::cout << "Current FPS: " << service.fps() << std::endl;
    std::cout << "Average latency: " << service.average_latency() << " ms" << std::endl;
    std::cout << "Min latency: " << service.min_latency() << " ms" << std::endl;
    std::cout << "Max latency: " << service.max_latency() << " ms" << std::endl;
    std::cout << "High latency alerts: " << service.alert_count() << std::endl;
}

void print_best_practices() {
    std::cout << "\n" << std::string(60, '=') << std::endl;
    std::cout << "Backend Best Practices:" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "1. Wrap model in service class for monitoring" << std::endl;
    std::cout << "2. Use sliding window for stable FPS calculation" << std::endl;
    std::cout << "3. Track latency statistics (min/max/avg)" << std::endl;
    std::cout << "4. Implement alerting for SLA compliance" << std::endl;
    std::cout << "5. Use warmup before serving production traffic" << std::endl;
    std::cout << "6. For advanced callbacks, use Python API" << std::endl;
}

void print_usage(const char* program) {
    std::cout << "Usage: " << program << " <model_path> [options]" << std::endl;
    std::cout << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  --device <device>  Target device (auto, cpu, cuda:0)" << std::endl;
    std::cout << "  --demo             Run demo inferences" << std::endl;
    std::cout << std::endl;
    std::cout << "Example:" << std::endl;
    std::cout << "  " << program << " yolov8n.onnx --device cuda:0 --demo" << std::endl;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        print_usage(argv[0]);
        return 1;
    }

    std::string model_path = argv[1];
    std::string device = "auto";
    bool run_demo_mode = false;

    // Parse arguments
    for (int i = 2; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--device" && i + 1 < argc) {
            device = argv[++i];
        } else if (arg == "--demo") {
            run_demo_mode = true;
        }
    }

    std::cout << std::string(60, '=') << std::endl;
    std::cout << "iVIT-SDK Backend Engineer Service Example" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "SDK Version: " << ivit::version() << std::endl;

    // Select device
    DeviceInfo target_device;
    if (device == "auto") {
        target_device = get_best_device();
        std::cout << "Auto-selected device: " << target_device.id
                  << " (" << target_device.name << ")" << std::endl;
    } else {
        target_device = DeviceManager::instance().get_device(device);
        std::cout << "Using device: " << target_device.id << std::endl;
    }

    // Step 1: Demonstrate monitoring patterns
    demonstrate_monitoring_patterns();

    // Create service
    std::cout << "\nCreating inference service..." << std::endl;
    InferenceService service(model_path, target_device);
    std::cout << "Service created on device: " << service.device_id()
              << " (" << service.device_name() << ")" << std::endl;

    // Step 2: Run demo
    if (run_demo_mode) {
        run_demo(service);
    }

    // Step 3: CLI tools
    demonstrate_cli_tools();

    // Best practices
    print_best_practices();

    return 0;
}

File: examples/python/advanced/backend_service.py
#!/usr/bin/env python3
"""
iVIT-SDK Backend Engineer Service Example

Target: Backend engineers who need to build stable AI inference services
        with monitoring and high concurrency support.

Features demonstrated:
- Callback system for monitoring
- FPS and latency tracking
- REST API service (FastAPI)
- Health checks and metrics

Usage:
    # Run standalone demo
    python backend_service.py

    # Run as REST API service
    python backend_service.py --serve --port 8080

    # With custom model
    python backend_service.py --serve --model path/to/model.onnx
"""

import argparse
import sys
import time
from typing import Optional

import ivit
from ivit.core.callbacks import (
    CallbackManager,
    CallbackContext,
    FPSCounter,
    LatencyLogger,
)


class PrometheusMetricsCallback:
    """Custom callback for Prometheus metrics integration."""

    def __init__(self):
        self.inference_count = 0
        self.total_latency = 0
        self.error_count = 0

    def __call__(self, ctx: CallbackContext):
        self.inference_count += 1
        self.total_latency += ctx.latency_ms

        # In production, send to Prometheus:
        # prometheus_client.Counter('inference_total').inc()
        # prometheus_client.Histogram('inference_latency').observe(ctx.latency_ms)

    @property
    def average_latency(self):
        if self.inference_count == 0:
            return 0
        return self.total_latency / self.inference_count


class AlertCallback:
    """Custom callback for alerting on high latency."""

    def __init__(self, threshold_ms: float = 100):
        self.threshold_ms = threshold_ms
        self.alert_count = 0

    def __call__(self, ctx: CallbackContext):
        if ctx.latency_ms > self.threshold_ms:
            self.alert_count += 1
            print(f"[ALERT] High latency warning: {ctx.latency_ms:.1f}ms "
                  f"(threshold: {self.threshold_ms}ms)")
            # In production: send Slack/Email notification


def setup_callbacks():
    """Step 1: Setup monitoring callbacks."""
    print("=" * 60)
    print("Step 1: Callback System Setup")
    print("=" * 60)

    callback_manager = CallbackManager()

    # Built-in: FPS Counter
    fps_counter = FPSCounter(window_size=30)
    callback_manager.register("infer_end", fps_counter)
    print("Registered: FPSCounter (window_size=30)")

    # Built-in: Latency Logger
    latency_logger = LatencyLogger()
    callback_manager.register("infer_end", latency_logger)
    print("Registered: LatencyLogger")

    # Custom: Prometheus Metrics
    prometheus_callback = PrometheusMetricsCallback()
    callback_manager.register("infer_end", prometheus_callback)
    print("Registered: PrometheusMetricsCallback")

    # Custom: Alert System
    alert_callback = AlertCallback(threshold_ms=100)
    callback_manager.register("infer_end", alert_callback)
    print("Registered: AlertCallback (threshold=100ms)")

    print("\nAvailable callback events:")
    print("  pre_process    - Before preprocessing")
    print("  post_process   - After postprocessing")
    print("  infer_start    - Inference starts")
    print("  infer_end      - Inference ends (with latency)")
    print("  batch_start    - Batch processing starts")
    print("  batch_end      - Batch processing ends")
    print("  stream_start   - Stream processing starts")
    print("  stream_frame   - Each frame in stream")
    print("  stream_end     - Stream processing ends")
    print("  model_loaded   - Model loaded")
    print("  model_unloaded - Model unloaded")

    return callback_manager, fps_counter, latency_logger, prometheus_callback, alert_callback


def inference_with_monitoring(
    model,
    image,
    callback_manager,
    fps_counter,
    prometheus_callback,
):
    """Step 2: Run inference with monitoring."""
    # Trigger infer_start event
    ctx = CallbackContext(event="infer_start", model_name="yolov8n")
    callback_manager.trigger("infer_start", ctx)

    # Execute inference
    start = time.perf_counter()
    results = model(image)
    latency = (time.perf_counter() - start) * 1000

    # Trigger infer_end event
    ctx = CallbackContext(
        event="infer_end",
        model_name="yolov8n",
        latency_ms=latency,
        detections=len(results)
    )
    callback_manager.trigger("infer_end", ctx)

    return results


def demonstrate_cli_tools():
    """Step 3: Demonstrate CLI tools for backend operations."""
    print("\n" + "=" * 60)
    print("Step 3: CLI Tools")
    print("=" * 60)

    print("Available CLI commands for backend engineers:")
    print()
    print("# System information")
    print("  ivit info")
    print()
    print("# List available devices")
    print("  ivit devices")
    print()
    print("# Performance benchmark")
    print("  ivit benchmark model.onnx --device cuda:0 --iterations 100")
    print()
    print("# Run inference")
    print("  ivit predict model.onnx image.jpg --output result.jpg")
    print()
    print("# Model conversion")
    print("  ivit convert model.onnx model.engine --format tensorrt --fp16")
    print()
    print("# Start REST API service")
    print("  ivit serve model.onnx --port 8080 --device cuda:0")
    print()
    print("# Model Zoo operations")
    print("  ivit zoo list")
    print("  ivit zoo search yolo")
    print("  ivit zoo download yolov8n")


def run_demo():
    """Run standalone demo without REST API."""
    print("=" * 60)
    print("iVIT-SDK Backend Engineer Service Demo")
    print("=" * 60)
    print(f"SDK Version: {ivit.__version__}")

    # Setup callbacks
    (callback_manager, fps_counter, latency_logger,
     prometheus_callback, alert_callback) = setup_callbacks()

    # Load model
    print("\n" + "=" * 60)
    print("Step 2: Model Loading")
    print("=" * 60)

    device = ivit.devices.best()
    print(f"Selected device: {device.id} ({device.name})")

    model = ivit.zoo.load("yolov8n", device=device)
    model.warmup(10)
    print("Model loaded and warmed up!")

    # Run demo inferences
    print("\n" + "=" * 60)
    print("Step 3: Demo Inferences")
    print("=" * 60)

    import numpy as np
    test_image = np.random.randint(0, 255, (480, 640, 3), dtype=np.uint8)

    print("Running 20 demo inferences...")
    for i in range(20):
        results = inference_with_monitoring(
            model, test_image,
            callback_manager, fps_counter, prometheus_callback
        )
        if (i + 1) % 5 == 0:
            print(f"  Completed {i + 1}/20, Current FPS: {fps_counter.fps:.1f}")

    # Show statistics
    print("\n" + "=" * 60)
    print("Step 4: Statistics")
    print("=" * 60)
    print(f"Total inferences: {prometheus_callback.inference_count}")
    print(f"Current FPS: {fps_counter.fps:.1f}")
    print(f"Average latency: {prometheus_callback.average_latency:.1f} ms")
    print(f"High latency alerts: {alert_callback.alert_count}")

    # Show CLI tools
    demonstrate_cli_tools()

    print("\n" + "=" * 60)
    print("Backend Best Practices:")
    print("=" * 60)
    print("1. Use callbacks for monitoring without modifying core logic")
    print("2. Implement health checks for load balancers")
    print("3. Track FPS and latency for SLA compliance")
    print("4. Use warmup before serving production traffic")
    print("5. Consider connection pooling for high concurrency")


def create_fastapi_app(model_path: Optional[str] = None):
    """Create FastAPI application for REST API service."""
    try:
        from fastapi import FastAPI, File, UploadFile, HTTPException
        from fastapi.responses import JSONResponse
        import numpy as np
        import cv2
    except ImportError:
        print("Error: FastAPI not installed.")
        print("Install with: pip install fastapi uvicorn python-multipart")
        sys.exit(1)

    app = FastAPI(
        title="iVIT Inference Service",
        description="REST API for AI inference using iVIT-SDK",
        version=ivit.__version__,
    )

    # Global state
    state = {
        "model": None,
        "fps_counter": None,
        "latency_logger": None,
        "inference_count": 0,
    }

    @app.on_event("startup")
    async def startup():
        """Initialize model on startup."""
        print("Starting iVIT Inference Service...")

        # Select best device
        device = ivit.devices.best()
        print(f"Selected device: {device.id} ({device.name})")

        # Load model
        if model_path:
            state["model"] = ivit.load(model_path, device=device)
        else:
            state["model"] = ivit.zoo.load("yolov8n", device=device)

        # Warmup
        state["model"].warmup(10)

        # Setup callbacks
        state["fps_counter"] = FPSCounter(window_size=100)
        state["latency_logger"] = LatencyLogger()
        state["model"].on("infer_end", state["fps_counter"])
        state["model"].on("infer_end", state["latency_logger"])

        print(f"Model loaded to {state['model'].device}")

    @app.post("/predict")
    async def predict(file: UploadFile = File(...)):
        """Execute object detection on uploaded image."""
        try:
            # Read image
            contents = await file.read()
            nparr = np.frombuffer(contents, np.uint8)
            image = cv2.imdecode(nparr, cv2.IMREAD_COLOR)

            if image is None:
                raise HTTPException(status_code=400, detail="Invalid image file")

            # Inference
            results = state["model"](image)
            state["inference_count"] += 1

            return JSONResponse({
                "success": True,
                "detections": results.to_dict().get("detections", []),
                "inference_time_ms": results.inference_time_ms,
                "current_fps": state["fps_counter"].fps,
            })

        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @app.get("/health")
    async def health():
        """Health check endpoint."""
        return {
            "status": "healthy",
            "model_loaded": state["model"] is not None,
            "device": str(state["model"].device) if state["model"] else None,
        }

    @app.get("/stats")
    async def stats():
        """Performance statistics endpoint."""
        return {
            "inference_count": state["inference_count"],
            "current_fps": state["fps_counter"].fps if state["fps_counter"] else 0,
            "average_latency_ms": (
                state["latency_logger"].average_latency
                if state["latency_logger"] else 0
            ),
            "device": str(state["model"].device) if state["model"] else None,
        }

    @app.get("/devices")
    async def devices():
        """List available devices."""
        device_list = []
        for device in ivit.devices():
            device_list.append({
                "id": device.id,
                "name": device.name,
                "backend": device.backend,
                "type": device.type,
            })
        return {"devices": device_list}

    return app


def run_server(port: int, model_path: Optional[str] = None):
    """Run FastAPI server."""
    try:
        import uvicorn
    except ImportError:
        print("Error: uvicorn not installed.")
        print("Install with: pip install uvicorn")
        sys.exit(1)

    app = create_fastapi_app(model_path)

    print("=" * 60)
    print("iVIT Inference Service")
    print("=" * 60)
    print(f"SDK Version: {ivit.__version__}")
    print(f"Starting server on port {port}...")
    print()
    print("API Endpoints:")
    print(f"  POST /predict - Run inference on uploaded image")
    print(f"  GET  /health  - Health check")
    print(f"  GET  /stats   - Performance statistics")
    print(f"  GET  /devices - List available devices")
    print()
    print("Example usage:")
    print(f"  curl -X POST http://localhost:{port}/predict -F 'file=@image.jpg'")
    print(f"  curl http://localhost:{port}/health")
    print(f"  curl http://localhost:{port}/stats")
    print()

    uvicorn.run(app, host="0.0.0.0", port=port)


def main():
    parser = argparse.ArgumentParser(
        description="iVIT-SDK Backend Engineer Service Example"
    )
    parser.add_argument(
        "--serve",
        action="store_true",
        help="Run as REST API service"
    )
    parser.add_argument(
        "--port",
        type=int,
        default=8080,
        help="Server port (default: 8080)"
    )
    parser.add_argument(
        "--model",
        type=str,
        default=None,
        help="Path to model file. If not provided, uses Model Zoo."
    )
    args = parser.parse_args()

    if args.serve:
        run_server(args.port, args.model)
    else:
        run_demo()


if __name__ == "__main__":
    main()
